share/functions/umask.fish

```diff
@@ -1,117 +1,124 @@
+end
 
-                case 'g*'
-                    set idx 2
-                    set i (echo $i| cut -c 2-)
+# Indexed by base umask value to be modified. Each digit represents the new value when the
+# corresponding index value is not and'd with the base umask value.
+set __fish_umask_remove_table 0101010 2002200 2103210 4440000 4541010 6442200 6543210
 
-                case 'o*'
-                    set idx 3
-                    set i (echo $i| cut -c 2-)
+function __fish_umask_remove
+    set -l val $argv[1]
+    set -l to_remove $argv[2]
 
-                case 'a*'
-                    set idx 1 2 3
-                    set i (echo $i| cut -c 2-)
+    set -l mask_table 0000000
+    if test $val -gt 0
+        set mask_table $__fish_umask_remove_table[$val]
+    end
+    set -l new_vals (string split '' $mask_table)
+    echo $new_vals[$to_remove]
+end
 
-                case '*'
-                    set implicit_all 1
-                    set idx 1 2 3
-            end
+# Given a umask string, possibly in symbolic mode, return an octal value with leading zeros.
+# This function expects to be called with a single value.
+function __fish_umask_parse
+    # Test if already a valid octal mask. If so pad it with zeros and return it.
+    # Note that umask values are always base 8 so they don't require a leading zero.
+    if string match -qr '^0?[0-7]{1,3}$' -- $argv
+        string sub -s -4 0000$argv
+        return 0
+    end
 
-            switch $i
-                case '=*'
-                    set mode set
-                    set i (echo $i| cut -c 2-)
-
-                case '+*'
-                    set mode add
-                    set i (echo $i| cut -c 2-)
-
-                case '-*'
-                    set mode remove
-                    set i (echo $i| cut -c 2-)
-
-                case '*'
-                    if not count $implicit_all >/dev/null
-                        printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
-                        return
-                    end
-                    set mode set
-            end
+    # Test if argument is a valid symbolic mask. Note that the basic pattern allows one illegal
+    # pattern: who and perms without a mode such as "urw". We test for that below after splitting
+    # the rights using the pattern and testing for that invalid combination.
+    set -l basic_pattern '([ugoa]*)([=+-]?)([rwx]*)'
+    if not string match -qr "^$basic_pattern(,$basic_pattern)*\$" -- $argv
+        printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
+        return 1
+    end
 
-            if not echo $perm | __fish_sgrep -E '^(r|w|x)*$' >/dev/null
-                printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
-                return
-            end
+    # Insert inverted umask into res variable.
+    set -l tmp (string split '' $umask)
+    set -l res
+    set res[1] (math 7 - $tmp[2])
+    set res[2] (math 7 - $tmp[3])
+    set res[3] (math 7 - $tmp[4])
```

Comment:
Good point. I had already implemented the lookup tables and related functions to eliminate the perl invocations before I tackled this part of the code. I'll take another look at this logic.